openedx-unsupported/ecommerce
# [ENT-235] Ask for consent on manual code entry

**Description:** This pull request ensures that when a coupon code linked to an Enterprise Customer is entered manually in the cart, consent is obtained prior to enrollment taking effect.

**JIRA:** [ENT-235](https://openedx.atlassian.net/browse/ENT-235)

**Dependencies:** Full behavior depends on edx/edx-enterprise#68

**Merge deadline:** 27 March 2017

**Installation instructions:** 

Requires a fully-configured E-Commerce installation. Caveats include:

- A commerce configuration must be set up in LMS pointing to the Ecommerce instance
- `ecommerce_worker` must exist be a staff user
- `enterprise_worker` must exist and be a staff user
- In a devstack, the E-commerce config is broken by default. Specifically, in the `JWT_AUTH` key, the `JWT_ISSUERS key must be a list of objects, rather than a list of strings:

    ```yaml
    JWT_AUTH:
        JWT_ISSUERS:
            - AUDIENCE: lms-key
              ISSUER: http://127.0.0.1:8000/oauth2
              SECRET_KEY: lms-secret
            - AUDIENCE: lms-key
              ISSUER: ecommerce_worker
              SECRET_KEY: lms-secret
    ```
- The OAuth client for ecommerce must be assigned one of `ecommerce_worker` and `enterprise_worker` as the linked user.

Additional configuration:

- Use the branch of edx-platform referenced in [this pull request](https://github.com/edx/edx-platform/pull/14708), as it's linked to the proper branch of `edx-enterprise`.
- Enterprise customer with consent enabled and required on enrollment
- Course discovery API populated with course metadata, and pushed to ecommerce
- Premium course mode created in E-commerce for demo course
- Coupon created and linked to that enterprise customer and course
- After the fact, entitlement added to the enterprise customer for that coupon

**Testing instructions:**

1. In ecommerce, download and open the coupon report for the above created coupon
2. Create a new LMS user account and activate it (required)
3. Browse to the course page for the demo course you made a premium mode for in Enterprise
4. Click Enroll.
5. Select the premium course mode
6. Observe that you're at a checkout page
7. Enter the coupon code from the report you downloaded directly.
8. Observe that you're taken to a consent prompt, without the coupon immediately applied.
9. Observe that if you decline consent, you're returned to the checkout page, and an error message appears.
10. Re-enter the code, and this time, consent to data sharing.
11. Observe that you're immediately enrolled in the course in the premium mode and taken to your Dashboard.
12. Log in with a superuser and, in the Manage Learners pane for your Enterprise Customer, observe that the new user you created is now a linked Enterprise Learner.
